parlai/torchscript/tokenizer.py

```diff
@@ -0,0 +1,689 @@
+                continue
+
+            # Capture runs of space characters
+            last_space_idx = idx
+            while last_space_idx + 1 < len(text) and text[last_space_idx + 1].isspace():
+                last_space_idx += 1
+            tokens.append(text[idx : last_space_idx + 1])
+            idx = last_space_idx + 1
+        return tokens
+
+    def __init__(
+        self,
+        add_prefix_space: bool,
+        encoder: Dict[str, str],
+        byte_encoder: Dict[int, str],
+        fused_key_bpe_ranks: Dict[str, float],
+        special_tokens: List[str],
+    ):
+
+        self.add_prefix_space = add_prefix_space
+
+        self.encoder = encoder
+        self.decoder: Dict[str, str] = {}
+        if self.encoder is not None:
+            for k, v in self.encoder.items():
+                self.decoder[v] = k
+
+        self.byte_encoder = byte_encoder
+        self.byte_decoder: Dict[str, int] = {}
+        if self.byte_encoder is not None:
+            for k, v in self.byte_encoder.items():
+                self.byte_decoder[v] = k
+
+        self.bpe_ranks = fused_key_bpe_ranks
+
+        # special tokens
+        self._special_tokens: Dict[str, int] = {}
+        for st in special_tokens:
+            self._special_tokens[st] = 1
+
+    def encode(self, text: str) -> List[str]:
+        """
+        Tokenize text.
+
+        Checks for add_prefix_space; handles accordingly.
+
+        :param text:
+            text to tokenize
+
+        :return tokens:
+            A list of tokens
+        """
+        if self.add_prefix_space:
+            text = f" {text}"
+
+        # constants for readability
+        FINAL = 1
+        SPLITABLE = 0
+        pieces: List[Tuple[str, int]] = [(text, SPLITABLE)]
+
+        for special_token in self._special_tokens.keys():
+            i = 0
+            while i < len(pieces):
+                subtext, status = pieces[i]
+                if status == FINAL:
+                    i += 1
+                    continue
+                split = subtext.split(special_token)
+                if len(split) > 1:
+                    # special token detected, replace the chunk with small subchunks
+                    # split by the special token
+                    pieces.pop(i)
+                    for j, piece in enumerate(split):
+                        if j > 0:
+                            # add the special token as a delimiter
+                            pieces.insert(i + (2 * j) - 1, (special_token, FINAL))
+                        pieces.insert(i + (2 * j), (piece, SPLITABLE))
+                else:
+                    i += 1
+
+        output: List[str] = []
+        for piece, state in pieces:
+            if state is FINAL:
+                output.append(piece)
+            else:
+                output += self.helper_encode(piece)
+        text = "".join(output)
+
+        return output
+
+    def get_pairs(self, word: List[str]) -> List[Tuple[str, str]]:
```

Comment:
Hmm this looks to be a duplicate of `ScriptableSubwordBpeHelper`. Can we create a superclass for shared methods? Not sure if Torchscripting will play nice with that

Reply:
Yes. These are duplicates. Earlier, I was trying to make one helper base.  It seems that Torch Script is not supporting inheritance yet? https://github.com/pytorch/pytorch/issues/41417

Reply:
Oof, that's what I was afraid of =/ Okay, this looks fine to me - will let @stephenroller approve because I think you asked him an outstanding question